runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/PipelineNode.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.construction.graph;
+
+import com.google.auto.value.AutoValue;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PCollection;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PTransform;
+
+/**
+ * A graph node which contains some pipeline element.
+ */
+interface PipelineNode {
+  enum NodeType {
+    PCOLLECTION,
+    PTRANSFORM
+  }
+
+  static PTransformNode ptransform(String id, PTransform transform) {
+    return new AutoValue_PipelineNode_PTransformNode(id, transform);
+  }
+
+  static PCollectionNode pcollection(String id, PCollection collection) {
+    return new AutoValue_PipelineNode_PCollectionNode(id, collection);
+  }
+
+  NodeType getType();
+
+  default PCollectionNode asPCollectionNode() {
```

Comment:
You will find that you'll want many node types within the runner depending on how things are mutated and how relevant pieces of information are found and I don't know if you'll want asYYY for each one vs just using instanceof and/or casting.

Reply:
ACK. Removed, and casting directly.